Adding complex number support for all_to_all/scatter

marjanfariborz opened the pull request:
Summary: Modifying all_to_all and scatter to support complex numbers as well as float numbers.

Test Plan: buck run //caffe2/test/distributed:distributed_gloo_fork -- test_name --print-passing-details --run-disabled

Differential Revision: D29563938

wanchaol commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2179,6 +2179,8 @@ def scatter(tensor, scatter_list=None, src=0, group=None, async_op=False):
     """
     Scatters a list of tensors to all processes in a group.
 
+    Complex tensors are supported.
```
nit: can we make this right before the `Args` section? we don't want to separate the descriptions

marjanfariborz replied:
Sure thing.

wanchaol commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2453,13 +2459,35 @@ def all_to_all_single(
+        >>> dist.all_to_all_single(output, input)
+        >>> output
+        tensor([1+1j, 5+5j, 9+9j, 13+13j])    # Rank 0
+        tensor([2+2j, 6+6j, 10+10j, 14+14j])    # Rank 1
```
nit: lint the trailing comment

wanchaol commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2549,6 +2579,23 @@ def all_to_all(output_tensor_list, input_tensor_list, group=None, async_op=False
+        >>> dist.all_to_all(output, input)
+        >>> output
+        [tensor([1+1j]), tensor([5+5j]), tensor([9+9j]), tensor([13+13j])]    # Rank 0
+        [tensor([2+2j]), tensor([6+6j]), tensor([10+10j]), tensor([14+14j])]    # Rank 1
```
ditto: lint the comment

wanchaol commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2670,6 +2674,11 @@ def test_scatter(self):
+        @unittest.skipIf(BACKEND == "nccl", "Nccl does not support scatter")
+        def test_scatter_complex(self):
+            group, group_id, rank = self._init_global_test()
+            self._test_scatter_helper(group, group_id, rank, dtype=torch.cfloat)
```
what's the difference between `torch.cfloat` and `torch.complex64`?

marjanfariborz replied:
They are both representing 64-bit complex data type.

wanchaol commented:
Looks good overall 👍 , thanks for working on this! Just have some minor comments/questions. For CI errors, I checked them and they seems not related to your change, so we are good to ship once we address the comments.

marjanfariborz committed:
Adding complex number support for all_to_all/scatter (#61299)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/61299

Modifying all_to_all and scatter to support complex numbers as well as float numbers.

Test Plan: buck run //caffe2/test/distributed:distributed_gloo_fork -- test_name --print-passing-details --run-disabled

Differential Revision: D29563938

fbshipit-source-id: 0d3649e60abe4075f92ba2ee97528e9c803093e7

rohan-varma reviewed (approved):
Awesome job and very comprehensive testing! Please also get a stamp from @wanchaol
